### PR TITLE
Allow renaming of properties with attrs

### PR DESCRIPTION
**Pull request by sebastianmika**
There are scenarios where the unstructured data have properties that are forbidden keywords in python. In an `attrs` class this property needs to take a different name.

Example data: `{'class': 'not allowed'}`

The following `attrs` class is **invalid** - at least I wouldn't see how that can be made working in `attrs` nor `cattr`:

```python
@attr.s
class MyClass(object):
    class = cattr.ib(type=str)
```

Proposal: use `attrs` metadata and provide an extension to `cattr` that allows to remap the invalid name to something valid when structuring and vice-versa when unstructuring.

I added a test, - but the `hypothesis` business is beyond my skills - sorry.

This works for me and is a solution similar to that taken by other packages (e.g. `related`):

```python
import attr
import cattr


@attr.s
class MyClass(object):
    renamed = cattr.ib(src_key='class', type=str)


def test_structure_keywords():
    data = {'class': 'value'}
    obj = cattr.structure(data, MyClass)
    assert cattr.unstructure(obj) == data
```

### CODE REVIEW

File: src/cattr/converters.py [hunk 1]
```diff
@@ -433,3 +441,21 @@ def _get_dis_func(self, union):
                 "currently. Register a loads hook manually."
             )
         return create_uniq_field_dis_func(*union_types)
+
+    def ib(self, default=attr.NOTHING, validator=None, repr=True, cmp=True,
+           hash=None, init=True, convert=None, metadata=None, type=None,
+           converter=None, factory=None, kw_only=False, src_key=None):
```
Comment (madsmtm) [inline src/cattr/converters.py]: Instead of the name `src_key`, we could also use something else. `Related` uses [`key`](https://github.com/genomoncology/related/blob/master/tests/ex05_field_names/test_renamed.py). `Marshmallow` uses [`data_key`](https://marshmallow.readthedocs.io/en/3.0/quickstart.html#specifying-serialization-deserialization-keys).

**Comment by madsmtm** [comment]
I don't think you have to worry about the failing tests, they're due to a change in `hypothesis` [`v4.0.0`](https://hypothesis.readthedocs.io/en/latest/changes.html#v4-0-0)

File: src/cattr/converters.py [hunk 2]
```diff
@@ -27,13 +27,17 @@
...
 from .multistrategy_dispatch import MultiStrategyDispatch
+import attr
 
 
 NoneType = type(None)
 T = TypeVar("T")
 V = TypeVar("V")
 
 
+CATTRS_METADATA_KEY = 'cattrs_structure_key'
```
[the remark this answers is not in the document]
Comment (sebastianmika) [reply src/cattr/converters.py]: I tried to follow the advise made here: http://www.attrs.org/en/stable/extending.html. I agree that it looks weird in my code to have an `cattr.ib`.

An alternative would be to use a more distinctive name like in the referenced example, e.g. `cattr.ib_mapped(..., src_key='mapped')`.

Of course, also this would work

```python
from cattr import CATTRS_METADATA_KEY

attrs.ib(..., metadata={CATTRS_METADATA_KEY: 'mapped'})
```
But it looks very ugly - at least `CATTRS_METADATA_KEY` should be more self explanatory + shorter.

What do you think @Tinche? Being active on `attrs` itself you certainly have a strong opinion.

File: src/cattr/converters.py (hunk 1 above)
Comment (sebastianmika) [reply src/cattr/converters.py]: I am absolutely agnostic about the name. Maybe `key` would actually be a better choice as we are mapping from a dictionary `key` to the property and back to the `key`.

I'll change it to be that.

File: src/cattr/converters.py [hunk 3]
```diff
@@ -318,6 +318,10 @@ def structure_attrs_fromdict(self, obj, cl):
             if name[0] == "_":
                 name = name[1:]
 
+            if val is None and type_ is not None:
+                raise ValueError('property "{:s}" of type "{:s}" must not be None'.format(
+                    src_key, str(type_)))
+
```
Comment (madsmtm) [inline src/cattr/converters.py]: Would this fix #53? And would it handle `Optional[str]`? I think it should be a separate PR though, since there's a bit of uncertainty as to what the best approach is, see #26

**Comment by Tinche** [comment]
Hi,

thank you for working on cattrs. I think this is a very useful feature to have, but I would prefer a different direction.

I think renaming attributes should be part of the (de)serialization logic, not the class itself, so I don't really like having it as an attribute in attr.ib(), even though I know other libraries and languages might use this approach (for example, Go).

On the master branch I've pushed an experimental new feature: a way of automatically generating unstructuring functions for attrs classes, and overriding their attributes. Here's an example:

```
import attr
from cattr import Converter
from cattr.generation import make_dict_unstructure_fn, override

@attr.s
class TestClass:
    a: str = attr.ib(default="a")

c = Converter()
c.register_unstructure_hook(TestClass, make_dict_unstructure_fn(TestClass, c, a=override(omit_if_default=True)))

>>> c.unstructure(TestClass())
{}   # 'a' had its default value so it was omitted.
```

We could easily support something like this:
```
@attr.s
class TestClass:
    klass: str = attr.ib(default="a")

c.register_unstructure_hook(TestClass, make_dict_unstructure_fn(TestClass, c, a=override(rename="class")))
```

These generated functions should be significantly faster than generic ones, and can support various overrides. The idea is, down the line, to have the converters generate them automatically, while also being overridable. The API is also clunky, and I have an idea on improving it.

The feature is not done - I still need to finish tests, iron out the implementation (like add support for generating structuring functions too) and write the docs. However feel free to take a look. I think adding rename support would be pretty easy, and since this is serialization logic, this is the right place for it to live.

**Comment by madsmtm** [comment]
To me, speed doesn't really matter, but I like the fact that I could now have two different converters, that could structure a `TestClass` from two different data sources! Also, this would allow us to easily use these features on external `attrs` classes. 👍

Just for reference, this is similar to the way `dacite` does it, see [dacite#remapping](https://github.com/konradhalas/dacite#remapping)

**Comment by sebastianmika** [comment]
I care a lot for speed - actually speed is a reason I would need to look for some other solution than `cattr`. So: I love you proposal @Tinche and happily close this PR. 

Time permitting (and wrapping my head around what you do there) I will try to contribute my 5 Cents to add the renaming logic.
